NDTensors/src/fill.jl

```diff
@@ -0,0 +1,54 @@
+# This is a file to write generic fills for NDTensors.
+#  This includes random fills, zeros, ...
+
+function generic_randn(
+  StoreT::Type{<:Dense{ElT,DataT}}, dim::Integer=0
+) where {DataT<:AbstractArray,ElT}
+  @assert ElT == eltype(DataT)
+  data = generic_randn(DataT, dim)
+  StoreT = set_datatype(StoreT, typeof(data))
+  return StoreT(data)
+end
+
+function generic_randn(StoreT::Type{<:Dense{ElT}}, dim::Integer=0) where {ElT}
+  return generic_randn(default_storagetype(ElT), dim)
+end
+
+function generic_randn(StoreT::Type{<:Dense}, dim::Integer=0)
+  return generic_randn(default_storagetype(), dim)
+end
+
+function generic_randn(DataT::Type{<:AbstractArray}, dim::Integer=0)
+  DataT = NDTensors.set_eltype_if_unspecified(DataT)
+  data = similar(DataT, dim)
+  ElT = eltype(DataT)
+  for i in 1:length(data)
+    data[i] = randn(ElT)
+  end
+  return data
+end
+
+##TODO replace randn in ITensors with generic_randn
+## and replace zeros with generic_zeros
+function generic_zeros(
+  StoreT::Type{<:Dense{ElT,DataT}}, dim::Integer=0
+) where {DataT<:AbstractArray,ElT}
+  @assert ElT == eltype(DataT)
+  data = generic_zeros(DataT, dim)
+  StoreT = NDTensors.set_datatype(StoreT, typeof(data))
```

Comment:
I don't think you should need `NDTensors.` here.

Reply:
Also, I'm confused by this line, doesn't `StoreT` already have a data type of type `DataT`, which is the same type as `data`? Or is this catching some cases where `data` isn't quite the same as `DataT`?

Reply:
Its catching the case when they aren't quite the same. This problem case came when you provide a `SubArray{Vector...}}` When you call `generic_zeros` it returns a `Vector` because `generic_zeros` uses `similar` which will unwrap types like we talked about yesterday. So then `DataT` doesn't match `typeof(data)` and I assumed that we shouldn't have the wrapped type as the DataT so I "fix" the  problem here

Reply:
Ok I think that makes sense, it would be weird to want to make a wrapper array filled with zeros. We could consider making that case error since we wouldn't be taking the input storage type very literally (like we try to do with constructor calls) but unwrapping seems like a reasonable behavior in this case.